src/mavlink-router/endpoint.cpp

```diff
@@ -795,17 +843,21 @@ int UdpEndpoint::write_msg(const struct buffer *pbuf)
     };
 
     _stat.write.total++;
-    _stat.write.bytes += pbuf->len;
+    _stat.write.bytes += r;
 
-    /* Incomplete packet, we warn and discard the rest */
-    if (r != (ssize_t) pbuf->len) {
-        _incomplete_msgs++;
-        log_debug("Discarding packet, incomplete write %zd but len=%u", r, pbuf->len);
-    }
+    tx_buf.len = std::max(ssize_t(0), (ssize_t)tx_buf.len - r);
+    // memcpy isn't safe for overlapping regions
+    memmove(tx_buf.data, &tx_buf.data[r], tx_buf.len);
```

Comment:
partial sends of frames is not something that can practically happen -- it means either you are violating the MTU of the protocol (logic error, you should terminate program), or the kernel has run out of memory (then you would have worse problems, and you should also terminate program)

possibly partial sends would also violate the deadline of the data that has been buffered, so there is little sense

Reply:
I was worrying about if someone dumps a giant amount of UDP over the internal interface faster than the link layer can sink it, but I guess in that case dropping data is the right thing to do anyway.